libs/vm-modules/tests/unit/math.cpp

```diff
@@ -383,4 +392,111 @@ TEST_F(MathTests, tensor_set_from_string)
   EXPECT_TRUE(gt.AllClose(tensor->GetTensor()));
 }
 
+TEST_F(MathTests, tensor_setup_charge_estimator_min_charge_succeeds)
+{
+  static char const *tensor_setup_src = R"(
+    function main()
+      var tensor_shape = Array<UInt64>(2);
+      tensor_shape[0] = 5u64;
+      tensor_shape[1] = 2u64;
+      var x = Tensor(tensor_shape);
+    endfunction
+  )";
+
+  ASSERT_TRUE(toolkit.Compile(tensor_setup_src));
+  ASSERT_TRUE(toolkit.Run(nullptr, tensor_setup_charge_min)) << stdout.str();
+}
+
+TEST_F(MathTests, tensor_setup_charge_estimator_min_charge_fails)
+{
+  static char const *tensor_setup_src = R"(
+    function main()
+      var tensor_shape = Array<UInt64>(2);
+      tensor_shape[0] = 5u64;
+      tensor_shape[1] = 2u64;
+      var x = Tensor(tensor_shape);
+    endfunction
+  )";
+
+  ASSERT_TRUE(toolkit.Compile(tensor_setup_src));
+  ASSERT_FALSE(toolkit.Run(nullptr, tensor_setup_charge_min - 1)) << stdout.str();
+}
+
+TEST_F(MathTests, tensor_fill_estimator_succeeds)
+{
+  size_t       tensor_size = 10;
+  ChargeAmount charge_max  = tensor_setup_charge_min + 2 * fetch::vm::CHARGE_UNIT * tensor_size;
+
+  static char const *tensor_to_string_src = R"(
+    function main()
+      var tensor_shape = Array<UInt64>(2);
+      tensor_shape[0] = 5u64;
+      tensor_shape[1] = 2u64;
+      var x = Tensor(tensor_shape);
+      x.fill(2.0fp64);
+    endfunction
+  )";
+
+  ASSERT_TRUE(toolkit.Compile(tensor_to_string_src));
+  ASSERT_TRUE(toolkit.Run(nullptr, charge_max)) << stdout.str();
+}
+
+TEST_F(MathTests, tensor_fill_estimator_fails)
+{
+  size_t       tensor_size = 10;
+  ChargeAmount charge_max  = tensor_setup_charge_min + 2 * fetch::vm::CHARGE_UNIT * tensor_size;
+
+  static char const *tensor_to_string_src = R"(
+    function main()
+      var tensor_shape = Array<UInt64>(2);
+      tensor_shape[0] = 5u64;
+      tensor_shape[1] = 5u64;
```

Comment:
Looking at this code, I question if this is the right level to test this. What you are checking is that the fill estimator returns a higher value if you give it a tensor with more work to do. Why not test that directly by calling the estimator in the unit test? The only reason you need magic numbers like

```
  size_t       tensor_size = 10;
  ChargeAmount charge_max  = tensor_setup_charge_min + 2 * fetch::vm::CHARGE_UNIT * tensor_size;
```

is that this test is running more code than you are interested in, which obscures the pass/fail criteria. Try extracting the estimators to a new file and testing them directly. Also, consider that some estimators may be so simple that they don't need testing.